src/Gaufrette/Adapter/GoogleCloudClientStorage.php

```diff
@@ -0,0 +1,295 @@
+    private $metadata     = array();
+
+    /**
+     * @param StorageClient    $service    Authenticated storage client class
+     * @param string           $bucketName Name of the bucket
+     * @param array            $options    Options are: "directory" and "acl" (see https://cloud.google.com/storage/docs/access-control/lists)
+     */
+    public function __construct(StorageClient $storageClient, $bucketName, $options = array())
+    {
+        $this->storageClient = $storageClient;
+        $this->initBucket($bucketName);
+        $this->options = array_replace_recursive(
+            array(
+                'directory' => '',
+                'acl'       => array(),
+            ),
+            $options
+        );
+        $this->options['directory'] = rtrim($this->options['directory'], '/');
+    }
+
+    /**
+     * Get adapter options
+     *
+     * @return  array
+     */
+    public function getOptions()
+    {
+        return $this->options;
+    }
+
+    /**
+     * Set adapter options
+     *
+     * @param   array   $options
+     */
+    public function setOptions($options)
+    {
+        $this->options = array_replace($this->options, $options);
+    }
+
+    /**
+     * @return Bucket
+     */
+    public function getBucket()
+    {
+        return $this->bucket;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function read($key)
+    {
+        $object = $this->bucket->object($this->computePath($key));
+
+        try {
+            return $object->downloadAsString();
+        } catch (\Exception $e) {
+            if ($e instanceof NotFoundException) {
+                throw new FileNotFound($key);
+            }
+
+            throw StorageFailure::unexpectedFailure('read', ['key' => $key], $e);
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function write($key, $content)
+    {
+        $options = array(
+            'resumable'     => true,
+            'name'          => $this->computePath($key),
+        );
+
+        try {
+            $object = $this->bucket->upload(
+                $content,
+                $options
+            );
+
+            $this->setAcl($object);
+        } catch (\Exception $e) {
+            throw StorageFailure::unexpectedFailure('write', ['key' => $key, 'content' => $content], $e);
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function exists($key)
+    {
+        return $this->bucket->object($this->computePath($key))->exists();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function isDirectory($key)
+    {
+        return $this->exists($this->computePath(rtrim($key, '/')).'/');
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function listKeys($prefix = null)
+    {
+        // @FIXME : the list should not return the root directory in the keys
+        $keys = array();
+
+        foreach ($this->bucket->objects(array('prefix' => $this->computePath($prefix))) as $e) {
```

Comment:
Lots of parenthesis here, could you put `objects()` result in a variable please? And maybe switch to newer array notation